Add option to use custom dataLayer variable

eljass opened the pull request:
This pull request introduces the ability to change the name of the `dataLayer` used by GTM. The new feature allows passing an option called `dataLayerName`, which initializes a window variable that contains the dataLayer array and changes helper methods such as `push()`, `trackEvent()`, and `trackView()` to push data to this renamed dataLayer variable.

When fetching Tag Manager's script, this uses Google's recommendation to pass the dataLayerName as a query parameter.

Shinigami92 commented on package-lock.json:
```diff
@@ -0,0 +1,6406 @@
+{
```
This project uses pnpm, please delete this file because pnpm-lock.yaml is already used.

You might need to use `corepack enable` on your machine

Shinigami92 commented on package.json:
```diff
@@ -8,7 +8,7 @@
     "lint": "eslint --cache --cache-strategy content --report-unused-disable-directives .",
     "ts-check": "tsc",
-    "test": "vitest",
+    "test": "vitest --silent=false",
```
This change is not related to the feature request
please revert it

Shinigami92 commented on package.json:
```diff
@@ -62,7 +62,7 @@
     "tsup": "~7.2.0",
     "typescript": "~5.2.2",
-    "vitest": "~0.34.5"
+    "vitest": "^0.34.6"
```
This change is not related to the feature request
please revert it

Shinigami92 commented on package.json:
```diff
@@ -62,7 +62,7 @@
   },
   "packageManager": "pnpm@8.7.6"
-}
+}
```
please run `pnpm run format` so the project is formatted correctly

Shinigami92 commented on src/data-layer-object.ts:
```diff
@@ -1,5 +1,5 @@
 /**
- * Object within the `window.dataLayer`.
+ * Object within the `window[pluginOptions[dataLayerName]]`.
```
I think this change is not really relevant.
We should only document necessary parts with that custom change.

Shinigami92 commented on src/gtm-support.ts:
```diff
@@ -144,14 +145,18 @@ export class GtmSupport {
 
   /**
-   * Returns the `window.dataLayer` array if the script is running in browser context and the plugin is enabled,
+   * Returns the `window[pluginOptions[dataLayerName]]` array if the script is running in browser context and the plugin is enabled,
```
This is not really correct, I think it is more like `window[dataLayerName]`

Shinigami92 commented on src/gtm-support.ts:
```diff
@@ -144,14 +145,18 @@ export class GtmSupport {
    * otherwise `false`.
    *
-   * @returns The `window.dataLayer` if script is running in browser context and plugin is enabled, otherwise `false`.
+   * @returns The `window[pluginOptions[dataLayerName]]` if script is running in browser context and plugin is enabled, otherwise `false`.
```
same here

Shinigami92 commented on src/index.ts:
```diff
@@ -3,7 +3,14 @@ import type { DataLayerObject } from './data-layer-object';
 declare global {
   interface Window {
     /**
-     * `dataLayer` used by GTM.
+     * Dynamic `dataLayer` key used by GTM.
+     *
+     * @see [developers.google.com/tag-manager/devguide/datalayer#rename_the_data_layer](https://developers.google.com/tag-platform/tag-manager/datalayer#rename_the_data_layer)
+     */
+    [key: string]: any;
```
This is a really bad idea. Instead we should document how to add TS support for a custom key if really necessary/wanted.

Shinigami92 commented on src/options.ts:
```diff
@@ -90,4 +90,12 @@ export interface GtmSupportOptions {
+   * @example dataLayerName: 'dataLayerDemo'
+   *
+   * @default dataLayer
+   */
```
we could add an `@see https://developers.google.com/tag-platform/tag-manager/datalayer#rename_the_data_layer` to this JSDoc :thinking:

Shinigami92 commented on src/options.ts:
```diff
@@ -90,4 +90,12 @@ export interface GtmSupportOptions {
+   *
+   * @example dataLayerName: 'dataLayerDemo'
+   *
+   * @default dataLayer
```
```suggestion
   * @default 'dataLayer'
```

Shinigami92 commented on src/utils.ts:
```diff
@@ -58,6 +58,12 @@ export interface LoadScriptOptions {
+   * The GTM dataLayer name.
+   *
+   * @default 'dataLayer'
+   */
```
We can add that `@see` here as well.

Shinigami92 commented on src/utils.ts:
```diff
@@ -82,9 +88,10 @@ export function loadScript(
+  window[dataLayerName] = window[dataLayerName] ?? [];
 
-  window.dataLayer?.push({
+  window[dataLayerName]?.push({
```
mh :thinking: not really related to your MR, but I think we could safely use

```suggestion
  window[dataLayerName]!.push({
```

here

eljass replied:
Sorted this out by getting or initializing dataLayer in reusable function

Shinigami92 commented on src/utils.ts:
```diff
@@ -106,6 +113,7 @@ export function loadScript(
 
   const queryString: URLSearchParams = new URLSearchParams({
     id,
+    l: config.dataLayerName ?? 'dataLayer',
```
IMO we should add the `l` query param only when it is **not** equals `'dataLayer'`

eljass commented:
@Shinigami92 I refactored the code based on your reviews.

Those made sense, and I'm sorry for not noticing the use of `pnpm` and the fixes to take such a long time. The main issue I experienced with this refactor was the change in the global window declaration. We'll somehow need to handle the types for the window so we can use the dynamic key for the `dataLayer` name. The solution is to have an extended interface for the window, which is used only inside this module and not passed on to the end product. I am not sure if documentation for typing the window is in the scope of this module or change as if the end user is using typescript, they should handle the window's types by themselves.

Let me know if the changes make sense for you, as I did create a reusable function to get or construct the `window.dataLayer` array. Otherwise, there would be lot of repetition in defining window type to allow dynamic naiming of the `dataLayer`. I did also refactored tests to use `dataLayer` from `instance`, to avoid same repetition. I don't know if you had some other justifications to not do so previously?
